### PR TITLE
stop after cycle point: support offsets

### LINKED ISSUE #5939: config: support offsets in hold and stop after cycle points
The final cycle point configuration supports offsets e.g:

```ini
final cycle point = 2000
final cycle point = 2000 + P1Y  # i.e. 2001
final cycle point = +P1Y  # i.e. "initial cycle point" + P1Y
```

However, the stop and hold cycle points do not.

It would be nice if these configs could work in the same way.

> Context / why is this an 8.3.0 issue?
>
> A common pattern in some areas is to run a workflow until it stops. Then modify the final cycle point to extend the workflow and restart.
>
> This isn't an easy working pattern to port to Cylc 8 due to the difficulty of re-populating the start tasks. We have a plan to make this easier, see #5416, however, this working pattern is needlessly complex in the first place.
>
> Swap `final cycle point` for `stop after cycle point` and the workflow never "completes" so the pool never needs re-polulating. This is much nicer for Cylc (no discontinuity), the user (no extra commands) and provenance (direct continuation of previous run).

### PR DESCRIPTION
**Pull request by oliver-sanders**
* Closes #5939
* Support offsets (measured from the ICP) for the `stop after cycle point` to mirror the behaviour of the `final cycle point`.
* Add integration test to lock-down `stop after cycle point` interactions.

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [x] Tests are included (or explain why tests are not needed).
- [x] `CHANGES.md` entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at cylc/cylc-doc/pull/XXXX.
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

### CODE REVIEW

File: cylc/flow/parsec/validate.py [hunk 1]
```diff
@@ -699,6 +700,30 @@ class CylcConfigValidator(ParsecValidator):
                 '-0830': 'UTC minus 8 hours and 30 minutes.'
             }
         ),
+        V_CYCLE_POINT_WITH_OFFSETS: (
```
Comment (oliver-sanders) [inline cylc/flow/parsec/validate.py:703]: This is a new parsec validator type which does the same thing as `V_STRING`.

It exists purely for documentation purposes.

File: cylc/flow/task_pool.py [hunk 2]
```diff
@@ -1026,7 +1026,7 @@ def set_stop_point(self, stop_point: 'PointBase') -> bool:
             LOG.info(f"Stop point unchanged: {stop_point}")
             return False
 
-        LOG.info("Setting stop point: {stop_point}")
+        LOG.info(f"Setting stop point: {stop_point}")
```
Comment (oliver-sanders) [inline cylc/flow/task_pool.py:1029]: Minor bugfix.

File: tests/unit/test_config.py [hunk 3]
```diff
@@ -114,7 +114,7 @@ def test_xfunction_imports(
         """
         flow_file.write_text(flow_config)
         workflow_config = WorkflowConfig(
-            workflow="name_a_tree", fpath=flow_file, options=Mock(spec=[]),
+            workflow="name_a_tree", fpath=flow_file, options=SimpleNamespace(spec=[]),
```
Comment (oliver-sanders) [inline tests/unit/test_config.py:117]: These tests were written with `Mock`, but none of them were actually relying on mocking in any way so converted them to `SimpleNamespace` which does the same thing, but with less unnecessary extras.

**Review by wxtim — CHANGES_REQUESTED** [review]
You have test failures - should be trivial fixes, as I think both are caused by a change in error message which reflects this PR being OK.

File: tests/unit/test_config.py (hunk 3 above)
Comment (MetRonnie) [reply tests/unit/test_config.py:117]: In that case, might as well be

```suggestion
            workflow="name_a_tree", fpath=flow_file, options=None,
```

or at least need to remove the `spec` arg if using `SimpleNamespace`

```suggestion
            workflow="name_a_tree", fpath=flow_file, options=SimpleNamespace(),
```

(Btw `Mock` is intended be used for stubbing too, it doesn't necessarily have to be used as a full-blown mock - https://docs.python.org/3/library/unittest.mock.html#the-mock-class)

File: tests/unit/test_config.py [hunk 4]
```diff
@@ -692,24 +698,28 @@ def test_process_fcp(
...
         ),
         pytest.param(
             '1993', '1066', '1066', '1066', None,
-            id="From options"
+            id="stop-cp-and-cli-option"
         ),
         pytest.param(
             '1993', 'reload', '1993', None, None,
-            id="From cfg if --stopcp=reload on restart"
+            id="stop-cp-and-cli-reload-option"
```
Comment (MetRonnie) [inline tests/unit/test_config.py:713]: I think the previous ids were more descriptive

File: tests/unit/test_config.py [hunk 5]
```diff
@@ -148,7 +148,7 @@ def test_xfunction_import_error(self, mock_glbl_cfg, tmp_path):
             WorkflowConfig(
                 workflow="caiman_workflow",
                 fpath=flow_file,
-                options=Mock(spec=[])
+                options=SimpleNamespace(spec=[])
```
Comment (MetRonnie) [inline tests/unit/test_config.py:151]:
```suggestion
                options=SimpleNamespace()
```

File: tests/unit/test_config.py [hunk 6]
```diff
@@ -179,7 +179,7 @@ def test_xfunction_attribute_error(self, mock_glbl_cfg, tmp_path):
         flow_file.write_text(flow_config)
         with pytest.raises(XtriggerConfigError) as excinfo:
             WorkflowConfig(workflow="capybara_workflow", fpath=flow_file,
-                           options=Mock(spec=[]))
+                           options=SimpleNamespace(spec=[]))
```
Comment (MetRonnie) [inline tests/unit/test_config.py:182]:
```suggestion
                           options=SimpleNamespace()
```

File: tests/unit/test_config.py [hunk 7]
```diff
@@ -211,7 +211,7 @@ def test_xfunction_not_callable(self, mock_glbl_cfg, tmp_path):
             WorkflowConfig(
                 workflow="workflow_with_not_callable",
                 fpath=flow_file,
-                options=Mock(spec=[])
+                options=SimpleNamespace(spec=[])
```
Comment (MetRonnie) [inline tests/unit/test_config.py:214]:
```suggestion
                options=SimpleNamespace()
```

File: tests/unit/test_config.py [hunk 8]
```diff
@@ -1170,7 +1185,7 @@ def test_check_circular(opt, monkeypatch, caplog, tmp_flow_config):
     # ----- Setup -----
     caplog.set_level(logging.WARNING, CYLC_LOG)
 
-    options = Mock(spec=[], is_validate=True)
+    options = SimpleNamespace(spec=[], is_validate=True)
```
Comment (MetRonnie) [inline tests/unit/test_config.py:1188]:
```suggestion
    options = SimpleNamespace(is_validate=True)
```

File: tests/unit/test_config.py [hunk 9]
```diff
@@ -1739,7 +1754,7 @@ def test_cylc_env_at_parsing(
 
     # Parse the workflow config then check the environment.
     WorkflowConfig(
-        workflow="name", fpath=flow_file, options=Mock(spec=[]),
+        workflow="name", fpath=flow_file, options=SimpleNamespace(spec=[]),
```
Comment (MetRonnie) [inline tests/unit/test_config.py:1757]:
```suggestion
        workflow="name", fpath=flow_file, options=SimpleNamespace(),
```

File: tests/unit/test_config.py (hunk 4 above)
Comment (oliver-sanders) [reply tests/unit/test_config.py:713]: fair enough, I changed them to something more CLI friendly, suggestions welcome.

Comment (MetRonnie) [reply tests/unit/test_config.py:713]: Having a short id plus a more descriptive comment would be reasonable IMO

**Pull request merged**